django/core/checks/translation.py

```diff
@@ -1,19 +1,78 @@
 from django.conf import settings
 from django.utils.translation.trans_real import language_code_re
 
-from . import Error, Tags, register
+from . import Error, Tags, Warning, register
 
 E001 = Error(
-    'You have provided an invalid value for the LANGUAGE_CODE setting.',
+    'You have provided an invalid value for the LANGUAGE_CODE setting: {}.',
     id='translation.E001',
 )
 
+E002 = Error(
+    'You have provided an invalid language code in the LANGUAGES setting: {}.',
+    id='translation.E002',
+)
+
+E003 = Error(
+    'You have provided an invalid language code in the LANGUAGES_BIDI setting: {}.',
+    id='translation.E003',
+)
+
+E004 = Error(
+    'You have provided a value for the LANGUAGE_CODE setting that is not in the LANGUAGES setting.',
+    id='translation.E004',
+)
+
+W005 = Warning(
+    'You have provided values in the LANGUAGES_BIDI setting that are not in the LANGUAGES setting.',
+    id='translation.W005',
+)
+
 
 @register(Tags.translation)
 def check_setting_language_code(app_configs, **kwargs):
     """
-    Errors if language code setting is invalid.
+    Errors if LANGUAGE_CODE setting is invalid.
     """
-    if not language_code_re.match(settings.LANGUAGE_CODE):
-        return [E001]
+    tag = settings.LANGUAGE_CODE
+    if not language_code_re.match(tag):
+        return [Error(E001.msg.format(tag), id=E001.id)]
     return []
+
+
+@register(Tags.translation)
+def check_setting_languages(app_configs, **kwargs):
+    """
+    Errors if LANGUAGES setting is invalid.
+    """
+    return [
+        Error(E002.msg.format(tag), id=E002.id)
+        for tag, _ in settings.LANGUAGES
+        if not language_code_re.match(tag)
```

Comment:
So to my understanding tags cannot be `None` and thus this would blow up. What if it were an integer or some other type? Would it be better to check `not isinstance(tag, str)` and use `{!r}` in the messages?

Reply:
Checking `isinstance(tag, str)` sounds good.